File: lw9/src/Image.ts
import { isPointInSize, Point, Size } from "./geom";
import {CoW} from "./CoW"
import {Tile} from "./Tile"

class Image {
	private _size: Size = new Size
	private _tiles: CoW<Tile>[][]

	constructor(size: Size, color: string) {
		if (size.width < 0 || size.height < 0) {
			throw new Error("Size cannot be less than 0")
		}

		if (color.length > 1) {
			throw new Error("Color must be a char")
		}
		const tileCountX = Math.ceil(size.width / Tile.SIZE);
        const tileCountY = Math.ceil(size.height / Tile.SIZE);

		this._size = size
		this._tiles = Array.from({ length: tileCountY }, () =>
            Array.from({ length: tileCountX }, () => new CoW(new Tile(color)))
        );
	}

	getSize(): Size {
		return this._size
	}

	getPixel(p: Point): string {
		if (p.x < 0 || p.y < 0 || p.x >= this._size.width || p.y >= this._size.height) {
            return ' ';
        }

        const tileX = Math.floor(p.x / Tile.SIZE);
        const tileY = Math.floor(p.y / Tile.SIZE);

        const localPoint: Point = {
            x: p.x % Tile.SIZE,
            y: p.y % Tile.SIZE
        };

        return this._tiles[tileY][tileX].getValue().getPixel(localPoint);
	}

	setPixel(p: Point, color: string): void {
		if (p.x < 0 || p.y < 0 || p.x >= this._size.width || p.y >= this._size.height) {
            return;
        }

        const tileX = Math.floor(p.x / Tile.SIZE);
        const tileY = Math.floor(p.y / Tile.SIZE);

        const localPoint: Point = {
            x: p.x % Tile.SIZE,
            y: p.y % Tile.SIZE
        };

        this._tiles[tileY][tileX].write(tile => {
            tile.setPixel(localPoint, color);
        });
	}
}

function printImage(image: Image): void {
    for (let y = 0; y < image.getSize().height; ++y) {
        let line = '';
        for (let x = 0; x < image.getSize().width; ++x) {
            line += image.getPixel({ x, y });
        }
        console.log(line);
    }
}

function loadImage(pixels: string): Image {
	const lines = pixels.split('\n');
    let maxWidth = 0;
    let height = lines.length;

    for (const line of lines) {
        maxWidth = Math.max(maxWidth, line.length);
    }

    const img = new Image({ width: maxWidth, height }, ' ');

    for (let y = 0; y < height; y++) {
        const line = lines[y];
        for (let x = 0; x < line.length; x++) {
            img.setPixel({ x, y }, line[x]);
        }
    }

    return img;
}

export {Image, printImage, loadImage}

File: lw9_c/src/image.cpp
#include "image.h"
#include <cassert>
#include <cmath>
#include <ostream>
#include <sstream>
#include <stdexcept>
#include <vector>

Image::Image(const Size size, const char color)
{
	if (size.width < 0 || size.height < 0)
	{
		throw std::out_of_range("Invalid size");
	}

	m_size = size;
	const int tileCountX = std::ceil(static_cast<double>(size.width) / static_cast<double>(Tile::SIZE));
	const int tileCountY = std::ceil(static_cast<double>(size.height) / static_cast<double>(Tile::SIZE));
	m_tiles.reserve(tileCountY);

	for (int y = 0; y < tileCountY; ++y)
	{
		m_tiles.emplace_back();
		m_tiles[y].reserve(tileCountX);
		for (int x = 0; x < tileCountX; ++x)
		{
			m_tiles[y].emplace_back(Tile(color));
		}
	}
}

Size Image::GetSize() const noexcept
{
	return m_size;
}

char Image::GetPixel(Point p) const noexcept
{
	if (!IsPointInSize(p, m_size))
	{
		return ' ';
	}

	const int tileX = p.x / Tile::SIZE;
	const int tileY = p.y / Tile::SIZE;

	const Point localPoint = {p.x % Tile::SIZE, p.y % Tile::SIZE};

	return m_tiles[tileY][tileX]->GetPixel(localPoint);
}

void Image::SetPixel(Point p, char color)
{
	if (!IsPointInSize(p, m_size))
	{
		return;
	}

	const int tileX = p.x / Tile::SIZE;
	const int tileY = p.y / Tile::SIZE;

	const Point localPoint = {p.x % Tile::SIZE, p.y % Tile::SIZE};

	m_tiles[tileY][tileX].Write([localPoint, color](Tile& tile)
	{
		tile.SetPixel(localPoint, color);
	});
}

void Print(const Image &img, std::ostream &out)
{
	const auto [width, height] = img.GetSize();
	for (int y = 0; y < height; ++y)
	{
		for (int x = 0; x < width; ++x)
		{
			out.put(img.GetPixel({x, y}));
		}
		out.put('\n');
	}
}

Image LoadImage(const std::string &pixels)
{
	std::istringstream s(pixels);
	Size size;
	std::string line;
	while (std::getline(s, line))
	{
		size.width = std::max(size.width, static_cast<int>(line.length()));
		++size.height;
	}

	Image img(size);

	s = std::istringstream(pixels);
	for (int y = 0; y < size.height; ++y)
	{
		if (!std::getline(s, line))
			break;

		int x = 0;
		for (const char ch: line)
		{
			img.SetPixel({x++, y}, ch);
		}
	}

	return img;
}
